### PR TITLE
fix dtype mismatch in lobpcg eigen solver

### LINKED ISSUE #132761: lobpcg doesn't support complex
### 🐛 Describe the bug

Calling torch.lobpcg with a torch.cfloat (or equivalent complex values) raise RuntimeError: "lt_cpu" not implemented for 'ComplexFloat'

when comparing the euclidian norm of covergence criteria equation 9 from https://arxiv.org/pdf/1704.07458 at line 792 from the code `converged = rerr < tol`

Since rerr is a complex type the comparision to a pure float is rightfully not implemented.

### Versions

torch==2.4.0

cc @albanD

### PR DESCRIPTION
**Pull request by sebastroy**
Fixes #132761

If rerr value is_complex, test against the real part. Since the rerr variable holds a norm calculation, the imaginary part will be 0.0.

### CODE REVIEW

File: torch/_lobpcg.py [hunk 1]
```diff
@@ -787,7 +787,10 @@ def update_converged_count(self):
             torch.norm(R, 2, (0,))
             * (torch.norm(X, 2, (0,)) * (A_norm + E[: X.shape[-1]] * B_norm)) ** -1
         )
-        converged = rerr < tol
+        if rerr.is_complex:
+            converged = rerr.real < tol  # this is a norm so imag is 0.0
+        else:
+            converged = rerr < tol
```
Comment (albanD) [inline torch/_lobpcg.py:793]:
```suggestion
        # For complex input, the result of the norm always has empty imaginary part
        converged = rerr.real < tol
```

For a real dtype, .real returns itself!

Comment (sebastroy) [reply torch/_lobpcg.py:793]: I tested it and got an error. Let me check again. Thanks!

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by hzhangxyz** [comment]
It seems we need to pass complex vectors X explicitly to lobpcg for complex matrix A, otherwise thw default X is still real vectors and then error raised.

**Pull request closed without merging**